python/whylogs/api/logger/experimental/logger/actor/process_rolling_logger.py

```diff
@@ -0,0 +1,473 @@
+        return [
+            Writers.get(
+                "whylabs",
+                # TODO get these from the session after that pr is merged
+                # org_id=self.env_vars.whylabs_org_id,
+                # api_key=self.env_vars.whylabs_api_key,
+                dataset_id=dataset_id,
+            )
+        ]
+
+    def _create_logger(self, dataset_id: str) -> ThreadRollingLogger:
+        writers = self.create_writers(dataset_id)
+
+        logger = ThreadRollingLogger(
+            aggregate_by=self.aggregate_by,
+            writers=[*self.debug_writers, *writers],
+            schema=self.schema,
+            write_schedule=self.write_schedule,
+            current_time_fn=self.current_time_ms,
+            queue_config=self._thread_queue_config,
+        )
+
+        self._logger.info(f"Created logger for {dataset_id}")
+        return logger
+
+    def _get_logger(self, dataset_id: str) -> ThreadRollingLogger:
+        if dataset_id not in self.loggers:
+            self.loggers[dataset_id] = self._create_logger(dataset_id)
+        return self.loggers[dataset_id]
+
+    def process_batch(self, batch: List[MessageType], batch_type: Type) -> None:
+        if batch_type == FlushMessage:
+            self.process_flush_message(cast(List[FlushMessage], batch))
+        elif batch_type == LogMessage:
+            self.process_log_messages(cast(List[LogMessage], batch))
+        elif batch_type == RawLogMessage:
+            self.process_raw_log_dicts(cast(List[RawLogMessage], batch))
+        elif batch_type == RawLogEmbeddingsMessage:
+            self.process_log_embeddings_messages(cast(List[RawLogEmbeddingsMessage], batch))
+        elif batch_type == RawPubSubMessage:
+            self.process_pubsub(cast(List[RawPubSubMessage], batch))
+        elif batch_type == RawPubSubEmbeddingMessage:
+            self.process_pubsub_embedding(cast(List[RawPubSubEmbeddingMessage], batch))
+        elif batch_type == CloseMessage:
+            self.process_close_message(cast(List[CloseMessage], batch))
+        elif batch_type == ProcessLoggerStatusMessage:
+            self._process_logger_status_message(cast(List[ProcessLoggerStatusMessage], batch))
+        else:
+            raise Exception(f"Unknown message type {batch_type}")
+
+    def process_close_message(self, messages: List[CloseMessage]) -> None:
+        self._logger.info("Running pre shutdown operations")
+        self._logger.info(f"Closing down {len(self.loggers)} loggers")
+        for datasetId, logger in self.loggers.items():
+            self._logger.info(f"Closing whylogs logger for {datasetId}")
+            logger.close()
+
+        if self._pipe_signaler is not None:
+            self._pipe_signaler.close_child()
+
+    def process_pubsub(self, messages: List[RawPubSubMessage]) -> None:
+        self._logger.info("Processing pubsub message")
+        msgs = [msg["log_request"] for msg in [it.to_pubsub_message() for it in messages] if msg is not None]
```

Comment:
Similar answer here to the previous one about get(). I'm using typed dicts for all of these types and the contract is that the type definitions are true, which I can enforce myself since these are all internal types that I make for myself.